### PR TITLE
r/glue_catalog_table - add `parameters` argument to `columns` + validations

### LINKED ISSUE #15778: Misleading documentation on aws_glue_catalog_table
The documentation for the aws_glue_catalog_table object blocks make it sound like it is a property excepting a list of strings or maps instead of a block.   The docs state: `partition_keys - (Optional) A list of columns by which the table is partitioned. Only primitive types are supported as partition keys.` but no reference to the object block that is allowed nor is there an example to demonstrate usage.  It seems that this property might have originally been a list of maps so maybe this doc string got missed.

I have referenced the s3_bucket cors_rule block as an example that is completely straightforward and seems to follow the convention used by other resources.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
$ terraform -v
Terraform v0.12.29
+ provider.aws v3.11.0
```

### Affected Resource(s)

* aws_glue_catalog_table

### References

* [glue_catalog_table#partition_keys](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/glue_catalog_table#partition_keys)
* [s3_bucket#cors_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#cors_rule)

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15778

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_glue_catalog_table - add support for `parameters` argument to `storage_descriptor.columns` block
resource_aws_glue_catalog_table - add plan time validation for `description`, `name`, `partition_keys.name`, `partition_keys.comment`, `partition_keys.type`, `retention`, `view_original_text`, `view_expanded_text`, `storage_descriptor.name`, `storage_descriptor.comment`, `storage_descriptor.type`, `storage_descriptor.bucket_columns`, `storage_descriptor.ser_de_info.name`, `storage_descriptor.skewed_info.skewed_column_names`, `storage_descriptor.sort_columns.column`, `storage_descriptor.sort_columns.sort_order`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSGlueCatalogTable_'
--- PASS: TestAccAWSGlueCatalogTable_disappears (42.12s)--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock (42.25s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock (43.59s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock (45.77s)
--- PASS: TestAccAWSGlueCatalogTable_columnParameters (48.80s)
--- PASS: TestAccAWSGlueCatalogTable_basic (52.32s)
--- PASS: TestAccAWSGlueCatalogTable_full (54.09s)
--- PASS: TestAccAWSGlueCatalogTable_recreates (59.63s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues (78.55s)
--- PASS: TestAccAWSGlueCatalogTable_update_addValues (79.21s)
--- PASS: TestAccAWSGlueCatalogTable_update_replaceValues (79.71s)
```

### CODE REVIEW

File: aws/resource_aws_glue_catalog_table.go [hunk 1]
```diff
@@ -39,13 +40,18 @@ func resourceAwsGlueCatalogTable() *schema.Resource {
...
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(0, 2048),
 			},
 			"name": {
 				Type:     schema.TypeString,
 				ForceNew: true,
 				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 255),
+					validation.StringDoesNotMatch(regexp.MustCompile(`[A-Z]`), "uppercase charcters cannot be used"),
```
Comment (ewbankkit) [inline aws/resource_aws_glue_catalog_table.go:53]:
```suggestion
					validation.StringDoesNotMatch(regexp.MustCompile(`[A-Z]`), "uppercase characters cannot be used"),
```

File: aws/resource_aws_glue_catalog_table.go [hunk 2]
```diff
@@ -243,9 +271,10 @@ func resourceAwsGlueCatalogTableCreate(d *schema.ResourceData, meta interface{})
 		TableInput:   expandGlueTableInput(d),
 	}
 
+	log.Printf("[DEBUG] Glue catalog table input: %#v", input)
 	_, err := conn.CreateTable(input)
 	if err != nil {
-		return fmt.Errorf("Error creating Catalog Table: %s", err)
+		return fmt.Errorf("Error creating Catalog Table: %w", err)
```
Comment (ewbankkit) [inline aws/resource_aws_glue_catalog_table.go:277]:
```suggestion
		return fmt.Errorf("Error creating Glue Catalog Table: %w", err)
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSGlueCatalogTable_' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=TestAccAWSGlueCatalogTable_ -timeout 120m
=== RUN   TestAccAWSGlueCatalogTable_recreates
=== PAUSE TestAccAWSGlueCatalogTable_recreates
=== RUN   TestAccAWSGlueCatalogTable_basic
=== PAUSE TestAccAWSGlueCatalogTable_basic
=== RUN   TestAccAWSGlueCatalogTable_columnParameters
=== PAUSE TestAccAWSGlueCatalogTable_columnParameters
=== RUN   TestAccAWSGlueCatalogTable_full
=== PAUSE TestAccAWSGlueCatalogTable_full
=== RUN   TestAccAWSGlueCatalogTable_update_addValues
=== PAUSE TestAccAWSGlueCatalogTable_update_addValues
=== RUN   TestAccAWSGlueCatalogTable_update_replaceValues
=== PAUSE TestAccAWSGlueCatalogTable_update_replaceValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_disappears
=== PAUSE TestAccAWSGlueCatalogTable_disappears
=== CONT  TestAccAWSGlueCatalogTable_recreates
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== CONT  TestAccAWSGlueCatalogTable_disappears
--- PASS: TestAccAWSGlueCatalogTable_disappears (14.93s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_recreates (21.19s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock (22.30s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock (13.38s)
=== CONT  TestAccAWSGlueCatalogTable_full
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock (13.86s)
=== CONT  TestAccAWSGlueCatalogTable_update_replaceValues
--- PASS: TestAccAWSGlueCatalogTable_full (16.79s)
=== CONT  TestAccAWSGlueCatalogTable_update_addValues
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues (38.56s)
=== CONT  TestAccAWSGlueCatalogTable_columnParameters
--- PASS: TestAccAWSGlueCatalogTable_update_replaceValues (26.23s)
=== CONT  TestAccAWSGlueCatalogTable_basic
--- PASS: TestAccAWSGlueCatalogTable_update_addValues (26.40s)
--- PASS: TestAccAWSGlueCatalogTable_columnParameters (15.11s)
--- PASS: TestAccAWSGlueCatalogTable_basic (15.02s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	77.463s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92 for the contribution!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSGlueCatalogTable_' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=TestAccAWSGlueCatalogTable_ -timeout 120m
=== RUN   TestAccAWSGlueCatalogTable_recreates
=== PAUSE TestAccAWSGlueCatalogTable_recreates
=== RUN   TestAccAWSGlueCatalogTable_basic
=== PAUSE TestAccAWSGlueCatalogTable_basic
=== RUN   TestAccAWSGlueCatalogTable_columnParameters
=== PAUSE TestAccAWSGlueCatalogTable_columnParameters
=== RUN   TestAccAWSGlueCatalogTable_full
=== PAUSE TestAccAWSGlueCatalogTable_full
=== RUN   TestAccAWSGlueCatalogTable_update_addValues
=== PAUSE TestAccAWSGlueCatalogTable_update_addValues
=== RUN   TestAccAWSGlueCatalogTable_update_replaceValues
=== PAUSE TestAccAWSGlueCatalogTable_update_replaceValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_disappears
=== PAUSE TestAccAWSGlueCatalogTable_disappears
=== CONT  TestAccAWSGlueCatalogTable_recreates
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== CONT  TestAccAWSGlueCatalogTable_full
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock (13.69s)
=== CONT  TestAccAWSGlueCatalogTable_update_replaceValues
--- PASS: TestAccAWSGlueCatalogTable_full (15.42s)
=== CONT  TestAccAWSGlueCatalogTable_update_addValues
--- PASS: TestAccAWSGlueCatalogTable_recreates (19.40s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock (13.42s)
=== CONT  TestAccAWSGlueCatalogTable_disappears
--- PASS: TestAccAWSGlueCatalogTable_update_replaceValues (25.62s)
=== CONT  TestAccAWSGlueCatalogTable_columnParameters
--- PASS: TestAccAWSGlueCatalogTable_update_addValues (26.43s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_disappears (12.02s)
=== CONT  TestAccAWSGlueCatalogTable_basic
--- PASS: TestAccAWSGlueCatalogTable_columnParameters (14.45s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock (12.91s)
--- PASS: TestAccAWSGlueCatalogTable_basic (13.85s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues (25.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	80.743s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSGlueCatalogTable_' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=TestAccAWSGlueCatalogTable_ -timeout 120m
=== RUN   TestAccAWSGlueCatalogTable_recreates
=== PAUSE TestAccAWSGlueCatalogTable_recreates
=== RUN   TestAccAWSGlueCatalogTable_basic
=== PAUSE TestAccAWSGlueCatalogTable_basic
=== RUN   TestAccAWSGlueCatalogTable_columnParameters
=== PAUSE TestAccAWSGlueCatalogTable_columnParameters
=== RUN   TestAccAWSGlueCatalogTable_full
=== PAUSE TestAccAWSGlueCatalogTable_full
=== RUN   TestAccAWSGlueCatalogTable_update_addValues
=== PAUSE TestAccAWSGlueCatalogTable_update_addValues
=== RUN   TestAccAWSGlueCatalogTable_update_replaceValues
=== PAUSE TestAccAWSGlueCatalogTable_update_replaceValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
=== RUN   TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== PAUSE TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
=== RUN   TestAccAWSGlueCatalogTable_disappears
=== PAUSE TestAccAWSGlueCatalogTable_disappears
=== CONT  TestAccAWSGlueCatalogTable_recreates
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock
=== CONT  TestAccAWSGlueCatalogTable_full
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_EmptyConfigurationBlock (10.37s)
=== CONT  TestAccAWSGlueCatalogTable_update_replaceValues
--- PASS: TestAccAWSGlueCatalogTable_full (11.99s)
=== CONT  TestAccAWSGlueCatalogTable_update_addValues
--- PASS: TestAccAWSGlueCatalogTable_recreates (14.40s)
=== CONT  TestAccAWSGlueCatalogTable_columnParameters
--- PASS: TestAccAWSGlueCatalogTable_columnParameters (10.48s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_update_replaceValues (18.05s)
=== CONT  TestAccAWSGlueCatalogTable_disappears
--- PASS: TestAccAWSGlueCatalogTable_update_addValues (17.96s)
=== CONT  TestAccAWSGlueCatalogTable_basic
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SkewedInfo_EmptyConfigurationBlock (10.17s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues
--- PASS: TestAccAWSGlueCatalogTable_disappears (8.49s)
=== CONT  TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock
--- PASS: TestAccAWSGlueCatalogTable_basic (11.62s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_EmptyConfigurationBlock (9.99s)
--- PASS: TestAccAWSGlueCatalogTable_StorageDescriptor_SerDeInfo_UpdateValues (16.96s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	53.873s
```

**Pull request merged**